### PR TITLE
Ignore records that are too big in Redshift destinations (instead of failing)

### LINKED ISSUE #2889: syncing to redshift fails if a row is too big
from https://airbytehq.slack.com/archives/C01MFR03D5W/p1618295157432200
## Expected Behavior
If a row that is being synced to redshift is too big then it should be dropped and the sync should not fail

## Current Behavior
The sync fails because the row is too large
Logs: [logs-28-0.txt](https://github.com/airbytehq/airbyte/files/6312817/logs-28-0.txt)
The user later communicated the error code they saw in the system table: 
```
err_code: 1204
err_reason: String length exceeds DDL length
```

This is happening while syncing the "Engagements" stream from Hubspot. 

## Severity of the bug for you
Medium

### PR DESCRIPTION
**Pull request by ChristopheDuong**
## What
Closes https://github.com/airbytehq/airbyte/issues/2889

## How

When an exception is raised by CopyConsumer class, the destination should now be able to abort the whole sync and terminates gracefully.

In case of a big record being sent to be loaded into redshift, we perform a size check pre-emptively and throws such records out before interacting with redshift.

This version of the fix would allow the sync to successfully completes for records that fits properly into redshift.

We could eventually iterate on another PR to better handle invalid records:
- upload them in a separate folder in s3 as rejected records?
- truncate json record to upload a smaller version of the data?
- another solution?

## Pre-merge Checklist
- [x] *Run integration tests*
- [x] *Publish Docker images*

### CODE REVIEW

**Comment by ChristopheDuong** [comment]
With this test case, I seemed to run into a hanging (infinitely?) on the `RedshiftCopyIntegrationTest`

But on the `RedshiftInsertIntegrationTest`, it fails with the following exceptions (not quite the same as the original issue):
```
2021-04-20 11:37:06 INFO i.a.c.i.LineGobbler(voidCall):69 - {} - 2021-04-20 11:37:06 INFO i.a.i.d.b.BufferedStreamConsumer(writeStreamsWithNRecords):195 - {} - Writing stream AirbyteStreamNameNamespacePair{name='withVeryLongRecords', namespace='null'}. Max batch size: 10000, Actual batch size: 1, Remaining buffered records: 0
2021-04-20 11:37:06 INFO i.a.c.i.LineGobbler(voidCall):69 - {} - 2021-04-20 11:37:06 INFO i.a.i.d.r.RedshiftSqlOperations(insertRecords):59 - {} - actual size of batch: 1
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - Exception in thread "main" java.lang.RuntimeException: java.sql.SQLException: [Amazon](500310) Invalid operation: Value too long for character type 
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - Details: -----------------------------------------------
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   error:  Value too long for character type
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   code:      8001
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   context:   Value too long for type character varying(65535)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   query:     3920277
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   location:  string.cpp:183
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   process:   padbmaster [pid=4155]
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   -----------------------------------------------;
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at io.airbyte.integrations.destination.buffered_stream_consumer.BufferedStreamConsumer.writeStreamsWithNRecords(BufferedStreamConsumer.java:199)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at io.airbyte.integrations.destination.buffered_stream_consumer.BufferedStreamConsumer.close(BufferedStreamConsumer.java:173)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at io.airbyte.integrations.base.FailureTrackingAirbyteMessageConsumer.close(FailureTrackingAirbyteMessageConsumer.java:82)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at io.airbyte.integrations.base.IntegrationRunner.consumeWriteStream(IntegrationRunner.java:138)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at io.airbyte.integrations.base.IntegrationRunner.run(IntegrationRunner.java:115)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at io.airbyte.integrations.destination.redshift.RedshiftDestination.main(RedshiftDestination.java:60)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - Caused by: java.sql.SQLException: [Amazon](500310) Invalid operation: Value too long for character type 
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - Details: -----------------------------------------------
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   error:  Value too long for character type
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   code:      8001
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   context:   Value too long for type character varying(65535)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   query:     3920277
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   location:  string.cpp:183
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   process:   padbmaster [pid=4155]
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   -----------------------------------------------;
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at com.amazon.redshift.client.messages.inbound.ErrorResponse.toErrorException(Unknown Source)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at com.amazon.redshift.client.PGMessagingContext.handleErrorResponse(Unknown Source)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at com.amazon.redshift.client.PGMessagingContext.handleMessage(Unknown Source)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at com.amazon.jdbc.communications.InboundMessagesPipeline.getNextMessageOfClass(Unknown Source)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at com.amazon.redshift.client.PGMessagingContext.doMoveToNextClass(Unknown Source)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at com.amazon.redshift.client.PGMessagingContext.getErrorResponse(Unknown Source)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at com.amazon.redshift.client.PGClient.handleErrorsScenario2ForPrepareExecution(Unknown Source)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at com.amazon.redshift.client.PGClient.handleErrorsPrepareExecute(Unknown Source)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at com.amazon.redshift.dataengine.CallablePreparedOrAtomicExecuteTask.call(Unknown Source)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at com.amazon.redshift.dataengine.CallablePreparedOrAtomicExecuteTask.call(Unknown Source)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - Caused by: com.amazon.support.exceptions.ErrorException: [Amazon](500310) Invalid operation: Value too long for character type 
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - Details: -----------------------------------------------
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   error:  Value too long for character type
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   code:      8001
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   context:   Value too long for type character varying(65535)
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   query:     3920277
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   location:  string.cpp:183
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   process:   padbmaster [pid=4155]
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} -   -----------------------------------------------;
2021-04-20 11:37:11 ERROR i.a.c.i.LineGobbler(voidCall):69 - {} - 	... 13 more

target process wasn't successful
io.airbyte.workers.WorkerException: target process wasn't successful
	at io.airbyte.workers.protocols.airbyte.DefaultAirbyteDestination.close(DefaultAirbyteDestination.java:107)
	at io.airbyte.integrations.standardtest.destination.TestDestination.runSync(TestDestination.java:603)
	at io.airbyte.integrations.standardtest.destination.TestDestination.testSync(TestDestination.java:307)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:628)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:117)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:184)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor$$Lambda$281/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:180)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:127)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$210/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$209/0000000000000000.invoke(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$208/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$DefaultDynamicTestExecutor.execute(NodeTestTask.java:198)
	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.execute(TestTemplateTestDescriptor.java:136)
	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.lambda$execute$2(TestTemplateTestDescriptor.java:104)
	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor$$Lambda$288/0000000000000000.accept(Unknown Source)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$11$1.accept(ReferencePipeline.java:442)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:274)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:497)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:487)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:239)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:274)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:497)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:487)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:239)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.execute(TestTemplateTestDescriptor.java:104)
	at org.junit.jupiter.engine.descriptor.TestTemplateTestDescriptor.execute(TestTemplateTestDescriptor.java:40)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$210/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$209/0000000000000000.invoke(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$208/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService$$Lambda$214/0000000000000000.accept(Unknown Source)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$210/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$209/0000000000000000.invoke(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$208/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService$$Lambda$214/0000000000000000.accept(Unknown Source)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$210/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$209/0000000000000000.invoke(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask$$Lambda$208/0000000000000000.execute(Unknown Source)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:248)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$5(DefaultLauncher.java:211)
	at org.junit.platform.launcher.core.DefaultLauncher$$Lambda$173/0000000000000000.accept(Unknown Source)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:226)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:199)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:132)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy4.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.stop(TestWorker.java:133)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:851)
```

**Review by cgardens — COMMENTED** [review]
interesting! is there any workaround for redshift for this to work that you can find? or do we need to something drastic like throw our records or truncate them?

File: airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/TestDestination.java [hunk 1]
```diff
@@ -495,6 +504,77 @@ public void testIncrementalDedupeSync() throws Exception {
...
+    final JsonNode config = getConfig();
+    final String defaultSchema = getDefaultSchema(config);
+    final List<AirbyteMessage> allMessages = new ArrayList<>();
+    allMessages.add(bigMessage);
+    allMessages.addAll(messages);
+    runSync(config, allMessages, configuredCatalog);
+    retrieveRawRecordsAndAssertSameMessages(catalog, messages, defaultSchema);
+  }
+
+  private String generateBigString(int addExtraCharacters) {
```
Comment (jrhizor) [inline airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/TestDestination.java:562]: nit: could use `RandomStringUtils` for this

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/CopyConsumer.java [hunk 2]
```diff
@@ -100,9 +104,16 @@ protected void acceptTracked(AirbyteRecordMessage message) throws Exception {
...
+    if (sqlOperations.isValidData(pair, data)) {
+      // TODO Truncate json data instead of throwing whole record away?
+      // or should we upload it into a special rejected record folder in s3 instead?
+      var emittedAt = Timestamp.from(Instant.ofEpochMilli(message.getEmittedAt()));
+      pairToCopier.get(pair).write(id, data, emittedAt);
+    }
+  }
 
-    pairToCopier.get(pair).write(id, data, emittedAt);
+  protected boolean isValidData(final AirbyteStreamNameNamespacePair streamName, final String data) {
```
Comment (jrhizor) [inline airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/CopyConsumer.java:115]: Is this used anywhere? Shouldn't we just be using the one in Sql operations?

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/CopyConsumer.java [hunk 3]
```diff
@@ -116,21 +127,30 @@ public void close(boolean hasFailed) throws Exception {
...
-        copier.closeStagingUploader(hasFailed);
-
-        if (!hasFailed) {
-          copier.createTemporaryTable();
-          copier.copyStagingFileToTemporaryTable();
-          copier.createDestinationSchema();
-          var destTableName = copier.createDestinationTable();
-          var mergeQuery = copier.generateMergeStatement(destTableName);
-          mergeCopiersToFinalTableQuery.append(mergeQuery);
+        try {
```
Comment (jrhizor) [inline airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/CopyConsumer.java:134]: Not blocking here, but generally it feels better to have visibility on an independent bug fix like this in the master commit messages.

"prevent snowflake s3 staging hanging on record-level failures" at that level would make it clear for someone running into this problem if/when a fix was introduced. They wouldn't be able to match to this title based on their problem.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftSqlOperations.java [hunk 4]
```diff
@@ -73,24 +75,14 @@ public void insertRecords(JdbcDatabase database, Stream<AirbyteRecordMessage> re
     SqlOperationsUtils.insertRawRecordsInSingleQuery(insertQueryComponent, recordQueryComponent, database, records);
   }
 
-  public void copyS3CsvFileIntoTable(JdbcDatabase database,
```
Comment (jrhizor) [inline airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftSqlOperations.java:76]: Where did this go?

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftSqlOperations.java [hunk 5]
```diff
@@ -73,24 +75,14 @@ public void insertRecords(JdbcDatabase database, Stream<AirbyteRecordMessage> re
...
-        s3FileLocation,
-        s3KeyId, s3Key,
-        s3Region);
-
-    database.execute(copyQuery);
+  @Override
+  public boolean isValidData(final AirbyteStreamNameNamespacePair streamName, final String data) {
+    final int dataSize = data.getBytes().length;
+    if (dataSize > REDSHIFT_VARCHAR_MAX_BYTE_SIZE) {
+      LOGGER.warn("Data({}) from stream {} exceeds limit of VARCHAR(65535) on Redshift... Ignoring record", dataSize, streamName);
```
Comment (jrhizor) [inline airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftSqlOperations.java:82]: I imagine usually if someone is running into this problem it's probably pervasive across that table. If it's a large table we don't want to log once per record.

I'm fine merging as is as long as we have a ticket to fix this problem.

**Review by jrhizor — APPROVED** [review]

**Comment by MC1-RicardoLeka** [comment]
instead off "Ignore records that are too big in Redshift destinations (instead of failing)" probably truncate the records and add a warning on the log.

**Comment by ChristopheDuong** [comment]
It's not so straightforward to truncate records since they are in JSON format and should remain as valid JSON objects, which is why I went with the ignore route instead for this PR...

Thus, truncation + warning or additional field signaling data was truncating should be contributed in another PR

**Pull request merged**